README.md
Mobile secure keyboard to prevent KeyLogger attack and screen capture.

## Screenshots
| Alphanumeric | Numeric |
|---|---|
| <img src="https://user-images.githubusercontent.com/47127353/103331966-8f5c8380-4aab-11eb-8098-f16c9417c2b7.png" width="200"> | <img src="https://user-images.githubusercontent.com/47127353/103331973-9a171880-4aab-11eb-8b34-fdf388d14044.png" width="200"> |

## Getting started

To use this plugin, add `flutter_secure_keyboard` as a [dependency in your pubspec.yaml file](https://flutter.io/platform-plugins/). For example:

```yaml
dependencies:
  flutter_secure_keyboard: ^1.0.5
```

## Examples

```dart
class WithSecureKeyboardExample extends StatefulWidget {
  @override
  _WithSecureKeyboardExampleState createState() => _WithSecureKeyboardExampleState();
}

class _WithSecureKeyboardExampleState extends State<WithSecureKeyboardExample> {
  final secureKeyboardController = SecureKeyboardController();

  final passwordEditor = TextEditingController();
  final passwordTextFieldFocusNode = FocusNode();

  final pinCodeEditor = TextEditingController();
  final pinCodeTextFieldFocusNode = FocusNode();

  @override
  Widget build(BuildContext context) {
    // We recommend that you set the secure keyboard to the top 
    // of the build function so that it can be seen properly.
    return WithSecureKeyboard(
      controller: secureKeyboardController,
      child: Scaffold(
        appBar: AppBar(title: Text('with_secure_keyboard_example')),
        body: _buildContentView()
      ),
    );
  }

  Widget _buildContentView() {
    // We recommend using the ListView widget to prevent widget overflow.
    return ListView(
      padding: const EdgeInsets.all(8.0),
      children: [
        Padding(
          padding: const EdgeInsets.only(bottom: 12.0),
          child: _buildPasswordTextField()
        ),
        _buildPinCodeTextField()
      ],
    );
  }

  Widget _buildPasswordTextField() {
    return Column(
      crossAxisAlignment: CrossAxisAlignment.start,
      children: [
        Text('Password'),
        TextFormField(
          controller: passwordEditor,
          focusNode: passwordTextFieldFocusNode,
          // We recommended to set false to prevent the soft keyboard from opening.
          enableInteractiveSelection: false,
          obscureText: true,
          onTap: () {
            secureKeyboardController.show(
              type: SecureKeyboardType.Alphanumeric,
              textFieldFocusNode: passwordTextFieldFocusNode,
              initText: passwordEditor.text,
              hintText: 'password',
              // Use onCharCodesChanged to have text entered in real time.
              onCharCodesChanged: (List<int> charCodes) {
                passwordEditor.text = String.fromCharCodes(charCodes);
              }
            );
          },
        ),
      ],
    );
  }

  Widget _buildPinCodeTextField() {
    return Column(
      crossAxisAlignment: CrossAxisAlignment.start,
      children: [
        Text('PinCode'),
        TextFormField(
          controller: pinCodeEditor,
          focusNode: pinCodeTextFieldFocusNode,
          // We recommended to set false to prevent the soft keyboard from opening.
          enableInteractiveSelection: false,
          obscureText: true,
          onTap: () {
            secureKeyboardController.show(
              type: SecureKeyboardType.Numeric,
              textFieldFocusNode: pinCodeTextFieldFocusNode,
              initText: pinCodeEditor.text,
              hintText: 'pinCode',
              // Use onDoneKeyPressed to allow text to be entered when you press the done key,
              // or to do something like encryption.
              onDoneKeyPressed: (List<int> charCodes) {
                pinCodeEditor.text = String.fromCharCodes(charCodes);
              }
            );
          },
        ),
      ],
    );
  }
}
```

## Package composition

* [WithSecureKeyboard] - A widget that implements a secure keyboard with controller.
* [SecureKeyboardController] - Controller to check or control the state of the secure keyboard.

### WithSecureKeyboard

| Parameter | Description |
|---|---|
| `controller`* | Controller for controlling the secure keyboard. |
| `child`* | A widget to have a secure keyboard. |
| `keyboardHeight` | Parameter to set the keyboard height. |
| `backgroundColor` | Parameter to set the keyboard background color. |
| `stringKeyColor` | Parameter to set keyboard string key(alphanumeric, numeric..) color. |
| `actionKeyColor` | Parameter to set keyboard action key(shift, backspace, clear..) color. |
| `doneKeyColor` | Parameter to set keyboard done key color. |
| `activatedKeyColor` | Set the color to display when activated with the shift action key. If the value is null, `doneKeyColor` is used. |
| `keyTextStyle` | Parameter to set keyboard key text style. |
| `inputTextStyle` | Parameter to set keyboard input text style. |
| `screenCaptureDetectedAlertTitle` | Security Alert title, only works on ios. |
| `screenCaptureDetectedAlertMessage` | Security Alert message, only works on ios |
| `screenCaptureDetectedAlertActionTitle` | Security Alert actionTitle, only works on ios. |

### SecureKeyboardController

| Function | Description |
|---|---|
| `isShowing` | Whether the secure keyboard is open. |
| `type` | Indicates the secure keyboard type. |
| `show` | Show a secure keyboard. |
| `hide` | Hide a secure keyboard. |

### SecureKeyboardController.show()

| Parameter | Description |
|---|---|
| `type`* | Specifies the secure keyboard type. |
| `textFieldFocusNode` | The `FocusNode` that will receive focus on. |
| `initText` | Set the initial value of the input text. |
| `hintText` | The hint text to display when the input text is empty. |
| `inputTextLengthSymbol` | Set the symbol to use when displaying the input text length. |
| `doneKeyText` | Set the done key text. |
| `clearKeyText` | Set the clear key text. |
| `obscuringCharacter` | Set the secure character to hide the input text. |
| `maxLength` | Set the maximum length of text that can be entered. |
| `alwaysCaps` | Whether to always display uppercase characters. |
| `obscureText` | Whether to hide input text as secure characters. |
| `onKeyPressed` | Called when the key is pressed. |
| `onCharCodesChanged` | Called when the character codes changed. |
| `onDoneKeyPressed` | Called when the done key is pressed. |
| `onCloseKeyPressed` | Called when the close key is pressed. |

CHANGELOG.md
## 1.0.6

* Example updates.
* Documentation updates.
* Add code so that the page does not close when the back button on the soft keyboard is pressed.

## 1.0.5

* README updates.
* Example updates.
* Remove onDoneKeyPressed required annotation.

## 1.0.4

* Add activatedKeyColor param.

## 1.0.3

* Change action key name from confirm to done.
* Fixed an error that occur when setting the key text.

## 1.0.2

* Add screenCaptureDetectedAlertTitle param.
* Add screenCaptureDetectedAlertMessage param.
* Add screenCaptureDetectedAlertActionTitle param.

## 1.0.1+3

* Update documentation.

## 1.0.1+2

* README updates.
* Change package name and structure.

## 1.0.1

* Fixed an issue that caused the input text to overflow.
* Added option to set maximum length of input text.

## 1.0.0+2

* Fixed an issue that did not work when dragging the view button.

## 1.0.0+1

* Resize the close and view buttons.

## 1.0.0

* Initial release.

lib/src/secure_keyboard.dart
import 'dart:async';
import 'dart:io';
import 'package:flutter/material.dart';
import 'package:flutter/services.dart';
import 'package:flutter_secure_keyboard/src/secure_keyboard_key_action.dart';
import 'package:flutter_secure_keyboard/src/secure_keyboard_key_generator.dart';
import 'package:flutter_secure_keyboard/src/secure_keyboard_key_type.dart';
import 'package:flutter_secure_keyboard/src/secure_keyboard_type.dart';
import 'package:flutter_secure_keyboard/src/secure_keyboard_key.dart';

/// Height of widget showing input text.
const double keyInputMonitorHeight = 50.0;

/// Keyboard default height.
const double keyboardDefaultHeight = 280.0;

/// Speed ​​of erasing input text when holding backspace.
const int backspaceEventDelay = 100;

/// Widget that implements a secure keyboard.
class SecureKeyboard extends StatefulWidget {
  /// Specifies the secure keyboard type.
  final SecureKeyboardType type;

  /// Called when the key is pressed.
  final ValueChanged<SecureKeyboardKey> onKeyPressed;

  /// Called when the character codes changed.
  final ValueChanged<List<int>> onCharCodesChanged;

  /// Called when the done key is pressed.
  final ValueChanged<List<int>> onDoneKeyPressed;

  /// Called when the close key is pressed.
  final VoidCallback onCloseKeyPressed;

  /// Set the initial value of the input text.
  final String initText;

  /// The hint text to display when the input text is empty.
  final String hintText;

  /// Set the symbol to use when displaying the input text length.
  final String inputTextLengthSymbol;

  /// Set the done key text.
  final String doneKeyText;

  /// Set the clear key text.
  final String clearKeyText;

  /// Set the secure character to hide the input text.
  final String obscuringCharacter;

  /// Set the maximum length of text that can be entered.
  final int maxLength;

  /// Whether to always display uppercase characters.
  final bool alwaysCaps;

  /// Whether to hide input text as secure characters.
  final bool obscureText;

  /// Parameter to set the keyboard height.
  final double height;

  /// Parameter to set the keyboard background color.
  final Color backgroundColor;

  /// Parameter to set keyboard string key(alphanumeric, numeric..) color.
  final Color stringKeyColor;

  /// Parameter to set keyboard action key(shift, backspace, clear..) color.
  final Color actionKeyColor;

  /// Parameter to set keyboard done key color.
  final Color doneKeyColor;

  /// Set the color to display when activated with the shift action key.
  /// If the value is null, `doneKeyColor` is used.
  final Color activatedKeyColor;

  /// Parameter to set keyboard key text style.
  final TextStyle keyTextStyle;

  /// Parameter to set keyboard input text style.
  final TextStyle inputTextStyle;

  /// Security Alert title, only works on ios.
  final String screenCaptureDetectedAlertTitle;

  /// Security Alert message, only works on ios.
  final String screenCaptureDetectedAlertMessage;

  /// Security Alert actionTitle, only works on ios.
  final String screenCaptureDetectedAlertActionTitle;

  SecureKeyboard({
    Key key,
    @required this.type,
    @required this.onKeyPressed,
    @required this.onCharCodesChanged,
    @required this.onDoneKeyPressed,
    @required this.onCloseKeyPressed,
    this.initText = '',
    this.hintText = '',
    this.inputTextLengthSymbol,
    this.doneKeyText,
    this.clearKeyText,
    this.obscuringCharacter = '•',
    this.maxLength,
    this.alwaysCaps = false,
    this.obscureText = true,
    this.height = keyboardDefaultHeight,
    this.backgroundColor = const Color(0xFF0A0A0A),
    this.stringKeyColor = const Color(0xFF313131),
    this.actionKeyColor = const Color(0xFF222222),
    this.doneKeyColor = const Color(0xFF1C7CDC),
    this.activatedKeyColor,
    this.keyTextStyle = const TextStyle(color: Colors.white, fontSize: 16.0, fontWeight: FontWeight.bold),
    this.inputTextStyle = const TextStyle(color: Colors.white, fontSize: 18.0, fontWeight: FontWeight.bold),
    this.screenCaptureDetectedAlertTitle,
    this.screenCaptureDetectedAlertMessage,
    this.screenCaptureDetectedAlertActionTitle
  })  : assert(type != null),
        assert(onKeyPressed != null),
        assert(onCharCodesChanged != null),
        assert(onDoneKeyPressed != null),
        assert(onCloseKeyPressed != null),
        assert(initText != null),
        assert(hintText != null),
        assert(obscuringCharacter != null && obscuringCharacter.isNotEmpty),
        assert(alwaysCaps != null),
        assert(obscureText != null),
        assert(height != null),
        assert(backgroundColor != null),
        assert(stringKeyColor != null),
        assert(actionKeyColor != null),
        assert(doneKeyColor != null),
        assert(keyTextStyle != null),
        assert(inputTextStyle != null),
        super(key: key);

  @override
  _SecureKeyboardState createState() => _SecureKeyboardState();
}

class _SecureKeyboardState extends State<SecureKeyboard> {
  final _methodChannel = const MethodChannel('flutter_secure_keyboard');

  final _definedKeyRows = List<List<SecureKeyboardKey>>();
  final _specialKeyRows = List<List<SecureKeyboardKey>>();
  final _charCodes = List<int>();
  
  Timer _backspaceEventGenerator;

  bool _isViewEnabled = false;
  bool _isShiftEnabled = false;
  bool _isSpecialCharsEnabled = false;
  
  void _initVariables() {
    _isViewEnabled = false;
    _isShiftEnabled = false;
    _isSpecialCharsEnabled = false;

    _definedKeyRows.clear();
    _specialKeyRows.clear();
    for (int i = 0; i < this._charCodes.length; i++) {
      this._charCodes[i] = 0x20;
    }
    this._charCodes.fillRange(0, this._charCodes.length, 0x20);
    this._charCodes.clear();
    this._charCodes.addAll(widget.initText.codeUnits);

    if (widget.type == SecureKeyboardType.Numeric)
      _definedKeyRows.addAll(SecureKeyboardKeyGenerator.instance.getNumericKeyRows());
    else
      _definedKeyRows.addAll(SecureKeyboardKeyGenerator.instance.getAlphanumericKeyRows());

    _specialKeyRows.addAll(SecureKeyboardKeyGenerator.instance.getSpecialCharsKeyRows());
  }

  void _onKeyPressed(SecureKeyboardKey key) {
    if (key.type == SecureKeyboardKeyType.String) {
      // The length of `charCodes` cannot exceed `maxLength`.
      if (widget.maxLength != null && widget.maxLength <= _charCodes.length)
        return;

      final keyText = (_isShiftEnabled || widget.alwaysCaps)
          ? key.capsText
          : key.text;
      setState(() => _charCodes.add(keyText.codeUnits.first));
      widget.onCharCodesChanged(_charCodes);
    } else if (key.type == SecureKeyboardKeyType.Action) {
      switch (key.action) {
        // Backspace
        case SecureKeyboardKeyAction.Backspace:
          if (_charCodes.isNotEmpty) {
            setState(() {
              this._charCodes.last = 0x20;
              this._charCodes.removeLast();
            });
            widget.onCharCodesChanged(_charCodes);
          }
          break;
          
        // Done
        case SecureKeyboardKeyAction.Done:
          widget.onDoneKeyPressed(_charCodes);
          break;
          
        // Clear
        case SecureKeyboardKeyAction.Clear:
          setState(() {
            for (int i = 0; i < this._charCodes.length; i++) {
              this._charCodes[i] = 0x20;
            }
            this._charCodes.fillRange(0, this._charCodes.length, 0x20);
            this._charCodes.clear();
          });
          widget.onCharCodesChanged(_charCodes);
          break;
          
        // Shift
        case SecureKeyboardKeyAction.Shift:
          if (!widget.alwaysCaps)
            setState(() {
              _isShiftEnabled = !_isShiftEnabled;
            });
          break;
          
        // SpecialChars
        case SecureKeyboardKeyAction.SpecialChars:
          setState(() {
            _isSpecialCharsEnabled = !_isSpecialCharsEnabled;
          });
          break;

        default:
          return;
      }
    }

    widget.onKeyPressed(key);
  }

  @override
  void didUpdateWidget(covariant SecureKeyboard oldWidget) {
    super.didUpdateWidget(oldWidget);
    _initVariables();
  }

  @override
  void initState() {
    super.initState();
    _methodChannel.invokeMethod('secureModeOn', {
      'screenCaptureDetectedAlertTitle': widget.screenCaptureDetectedAlertTitle,
      'screenCaptureDetectedAlertMessage': widget.screenCaptureDetectedAlertMessage,
      'screenCaptureDetectedAlertActionTitle': widget.screenCaptureDetectedAlertActionTitle
    });
    _initVariables();
  }

  @override
  Widget build(BuildContext context) {
    final keyRows = _isSpecialCharsEnabled
        ? _specialKeyRows
        : _definedKeyRows;
    final keyboardKey = _buildKeyboardKey(keyRows);
    keyboardKey.insert(0, _buildKeyInputMonitor());

    return WillPopScope(
      onWillPop: widget.onCloseKeyPressed,
      child: Container(
        width: MediaQuery.of(context).size.width,
        height: widget.height + keyInputMonitorHeight,
        color: widget.backgroundColor,
        child: Column(
          mainAxisAlignment: MainAxisAlignment.spaceBetween,
          crossAxisAlignment: CrossAxisAlignment.center,
          children: keyboardKey
        ),
      ),
    );
  }

  @override
  void dispose() {
    _methodChannel.invokeMethod('secureModeOff');
    for (int i = 0; i < this._charCodes.length; i++) {
      this._charCodes[i] = 0x20;
    }
    this._charCodes.fillRange(0, this._charCodes.length, 0x20);
    super.dispose();
  }

  Widget _buildKeyInputMonitor() {
    String secureText;
    TextStyle secureTextStyle;

    if (_charCodes.isNotEmpty) {
      secureText = '';
      for (var i=0; i<_charCodes.length; i++) {
        // if (i == _charCodes.length - 1)
        //   secureText += String.fromCharCode(_charCodes[i]);
        // else
        secureText += widget.obscuringCharacter;
      }

      secureTextStyle = widget.inputTextStyle;
    } else {
      secureText = widget.hintText;
      secureTextStyle = widget.inputTextStyle.copyWith(
          color: widget.inputTextStyle.color.withOpacity(0.5));
    }

    final lengthSymbol = widget.inputTextLengthSymbol ?? (Platform.localeName == 'ko_KR')
        ? '자'
        : 'digit';
    final lengthText = '${_charCodes.length}$lengthSymbol';

    Widget viewKey = SizedBox();
    if (widget.obscureText) {
      viewKey = Container(
        width: keyInputMonitorHeight / 1.4,
        height: keyInputMonitorHeight / 1.4,
        margin: const EdgeInsets.only(left: 1.5),
        child: Material(
          color: Colors.transparent,
          child: InkWell(
            borderRadius: BorderRadius.circular(24.0),
            onTap: () {

            },
            child: GestureDetector(
              onTapDown: (_) => setState(() => _isViewEnabled = true),
              onTapUp: (_) => setState(() => _isViewEnabled = false),
              onPanEnd: (_) => setState(() => _isViewEnabled = false),
              child: Icon(Icons.remove_red_eye, color: widget.keyTextStyle.color)
            )
          ),
        ),
      );
    }

    return SizedBox(
      height: keyInputMonitorHeight,
      child: Row(
        children: [
          Expanded(
            child: Padding(
              padding: const EdgeInsets.only(left: 8.0),
              child: Text(
                secureText,
                style: secureTextStyle,
                maxLines: 1,
                overflow: TextOverflow.ellipsis
              )
            )
          ),
          Padding(
            padding: const EdgeInsets.symmetric(horizontal: 5.0),
            child: Text(lengthText, style: widget.keyTextStyle)
          ),
          viewKey,
          Container(
            width: keyInputMonitorHeight / 1.4,
            height: keyInputMonitorHeight / 1.4,
            margin: const EdgeInsets.only(right: 1.5),
            child: Material(
              color: Colors.transparent,
              child: InkWell(
                borderRadius: BorderRadius.circular(24.0),
                onTap: widget.onCloseKeyPressed,
                child: Icon(Icons.close, color: widget.keyTextStyle.color)
              ),
            ),
          )
        ],
      ),
    );
  }

  List<Widget> _buildKeyboardKey(List<List<SecureKeyboardKey>> keyRows) {
    return List.generate(keyRows.length, (int rowNum) {
      return Material(
        color: Colors.transparent,
        child: Row(
          mainAxisAlignment: MainAxisAlignment.spaceAround,
          crossAxisAlignment: CrossAxisAlignment.center,
          children: List.generate(keyRows[rowNum].length, (int keyNum) {
            final key = keyRows[rowNum][keyNum];
            
            switch (key.type) {
              case SecureKeyboardKeyType.String:
                return _buildStringKey(key, keyRows.length);
              case SecureKeyboardKeyType.Action:
                return _buildActionKey(key, keyRows.length);
              default:
                throw Exception('Unknown key type.');
            }
          })
        ),
      );
    });
  }
  
  Widget _buildStringKey(SecureKeyboardKey key, int keyRowsLength) {
    final keyText = (_isShiftEnabled || widget.alwaysCaps)
        ? key.capsText
        : key.text;

    return Expanded(
      child: Container(
        height: widget.height / keyRowsLength,
        padding: const EdgeInsets.all(1.5),
        child: Material(
          borderRadius: BorderRadius.circular(4.0),
          color: widget.stringKeyColor,
          child: InkWell(
            onTap: () => _onKeyPressed(key),
            child: Center(child: Text(keyText, style: widget.keyTextStyle))
          ),
        ),
      ),
    );
  }

  Widget _buildActionKey(SecureKeyboardKey key, int keyRowsLength) {
    String keyText;
    Widget actionKey;

    switch (key.action) {
      case SecureKeyboardKeyAction.Backspace:
        actionKey = GestureDetector(
          onLongPress: () {
            final delay = Duration(milliseconds: backspaceEventDelay);
            _backspaceEventGenerator = Timer.periodic(delay, (_) => _onKeyPressed(key));
          },
          onLongPressUp: () {
            if (_backspaceEventGenerator != null) {
              _backspaceEventGenerator.cancel();
              _backspaceEventGenerator = null;
            }
          },
          child: Icon(Icons.backspace, color: widget.keyTextStyle.color)
        );
        break;
      case SecureKeyboardKeyAction.Shift:
        actionKey = Icon(Icons.arrow_upward, color: widget.keyTextStyle.color);
        break;
      case SecureKeyboardKeyAction.Clear:
        keyText = widget.clearKeyText;
        if (keyText == null || keyText.isEmpty)
          keyText = (Platform.localeName == 'ko_KR') ? '초기화' : 'Clear';

        actionKey = Text(keyText, style: widget.keyTextStyle);
        break;
      case SecureKeyboardKeyAction.Done:
        keyText = widget.doneKeyText;
        if (keyText == null || keyText.isEmpty)
          keyText = (Platform.localeName == 'ko_KR') ? '입력완료' : 'Done';

        actionKey = Text(keyText, style: widget.keyTextStyle);
        break;
      case SecureKeyboardKeyAction.SpecialChars:
        actionKey = Text(
          _isSpecialCharsEnabled
              ? (_isShiftEnabled ? 'ABC' : 'abc')
              : '!@#',
          style: widget.keyTextStyle
        );
        break;
      case SecureKeyboardKeyAction.Blank:
        return Expanded(child: SizedBox());
    }

    Color keyColor;
    if (key.action == SecureKeyboardKeyAction.Done)
      keyColor = widget.doneKeyColor;
    else if (key.action == SecureKeyboardKeyAction.Shift && _isShiftEnabled)
      keyColor = widget.activatedKeyColor ?? widget.doneKeyColor;
    else
      keyColor = widget.actionKeyColor;

    return Expanded(
      child: Container(
        height: widget.height / keyRowsLength,
        padding: const EdgeInsets.all(1.5),
        child: Material(
          borderRadius: BorderRadius.circular(4.0),
          color: keyColor,
          child: InkWell(
            onTap: () => _onKeyPressed(key),
            child: Center(child: actionKey)
          ),
        ),
      ),
    );
  }
}


lib/src/secure_keyboard_key.dart
import 'package:flutter/material.dart';
import 'package:flutter_secure_keyboard/src/secure_keyboard_key_action.dart';
import 'package:flutter_secure_keyboard/src/secure_keyboard_key_type.dart';

/// Class defining secure keyboard key.
class SecureKeyboardKey {
  /// Key text in lowercase form.
  final String text;

  /// Key text in uppercase form.
  final String capsText;

  /// Key type (Action, String)
  final SecureKeyboardKeyType type;

  /// Key action (Backspace, Confirm, Clear, Shift, Blank, SpecialChars)
  final SecureKeyboardKeyAction action;

  SecureKeyboardKey({
    this.text,
    this.capsText,
    @required this.type,
    this.action
  })  : assert(type != null);

  /// Returns the class field in map form.
  Map<String, dynamic> toMap() {
    return {
      'text': text,
      'capsText': capsText,
      'type': type,
      'action': action
    };
  }
}


lib/src/secure_keyboard_key_generator.dart
import 'dart:math';
import 'package:flutter_secure_keyboard/src/secure_keyboard_key.dart';
import 'package:flutter_secure_keyboard/src/secure_keyboard_key_action.dart';
import 'package:flutter_secure_keyboard/src/secure_keyboard_key_type.dart';

/// Class responsible for key generation of secure keyboard.
class SecureKeyboardKeyGenerator {
  SecureKeyboardKeyGenerator._internal();
  static final instance = SecureKeyboardKeyGenerator._internal();

  /// Maximum length of a row of numeric key.
  /// If not enough, fill in the blank action key.
  final int _numericKeyRowMaxLength = 4;
  final List<List<String>> _numericKeyRows = [
    const ['1', '2', '3'],
    const ['4', '5', '6'],
    const ['7', '8', '9', '0'],
    const []
  ];

  /// Maximum length of a row of alphanumeric key.
  /// If not enough, fill in the blank action key.
  final int _alphanumericKeyRowMaxLength = 11;
  final List<List<String>> _alphanumericKeyRows = [
    const ['1', '2', '3', '4', '5', '6', '7', '8', '9', '0'],
    const ['q', 'w', 'e', 'r', 't', 'y', 'u', 'i', 'o', 'p'],
    const ['a', 's', 'd', 'f', 'g', 'h', 'j', 'k', 'l'],
    const ['z', 'x', 'c', 'v', 'b', 'n', 'm'],
    const []
  ];

  /// Maximum length of a row of special characters key.
  /// If not enough, fill in the blank action key.
  final int _specialCharsKeyRowMaxLength = 11;
  final List<List<String>> _specialCharsKeyRows = [
    const ['!', '@', '#', '\$', '%', '^', '&', '*', '(', ')'],
    const ['-', '=', '+', '{', '}', '[', ']', '\\', ':', ';'],
    const ['\"', '\'', '<', '>', ',', '.', '/', '?', '|'],
    const ['~', '`', '_', '\₩', '#', '@', '!'],
    const []
  ];

  /// Returns a list of numeric key rows.
  List<List<SecureKeyboardKey>> getNumericKeyRows() {
    final random = Random();
    int randomIndex;
    int emptyLength;

    return List.generate(_numericKeyRows.length, (int rowNum) {
      List<SecureKeyboardKey> rowKeys = [];

      switch (rowNum) {
        case 3:
          // Clear
          rowKeys.add(_clearActionKey());

          // Backspace
          rowKeys.add(_backspaceActionKey());

          // Confirm
          rowKeys.add(_confirmActionKey());
          break;
        default:
          rowKeys = _getStringKeyRow(_numericKeyRows, rowNum);
          emptyLength = _numericKeyRowMaxLength - _numericKeyRows[rowNum].length;

          for (var i=0; i<emptyLength; i++) {
            randomIndex = random.nextInt(_numericKeyRowMaxLength);

            if (randomIndex == _numericKeyRowMaxLength - 1)
              rowKeys.add(_blankActionKey());
            else
              rowKeys.insert(randomIndex, _blankActionKey());
          }

          rowKeys.shuffle();
      }

      return rowKeys;
    });
  }

  /// Returns a list of alphanumeric key rows.
  List<List<SecureKeyboardKey>> getAlphanumericKeyRows() {
    final random = Random();
    int randomIndex;
    int emptyLength;

    return List.generate(_alphanumericKeyRows.length, (int rowNum) {
      List<SecureKeyboardKey> rowKeys = [];

      switch (rowNum) {
        case 3:
          // Shift
          rowKeys.add(_shiftActionKey());

          rowKeys.addAll(_getStringKeyRow(_alphanumericKeyRows, rowNum));

          emptyLength = _alphanumericKeyRows[rowNum].length;
          randomIndex = random.nextInt(emptyLength) + 1;
          rowKeys.insert(randomIndex, _blankActionKey());

          // Backspace
          rowKeys.add(_backspaceActionKey());
          break;
        case 4:
          // SpecialChars
          rowKeys.add(_specialCharsActionKey());

          // Clear
          rowKeys.add(_clearActionKey());

          // Confirm
          rowKeys.add(_confirmActionKey());
          break;
        default:
          rowKeys = _getStringKeyRow(_alphanumericKeyRows, rowNum);
          emptyLength = _alphanumericKeyRowMaxLength - _alphanumericKeyRows[rowNum].length;

          for (var i=0; i<emptyLength; i++) {
            randomIndex = random.nextInt(_alphanumericKeyRowMaxLength);

            if (randomIndex == _alphanumericKeyRowMaxLength - 1)
              rowKeys.add(_blankActionKey());
            else
              rowKeys.insert(randomIndex, _blankActionKey());
          }

          if (rowNum == 0)
            rowKeys.shuffle();
      }

      return rowKeys;
    });
  }

  /// Returns a list of special characters key rows.
  List<List<SecureKeyboardKey>> getSpecialCharsKeyRows() {
    final random = Random();
    int randomIndex;
    int emptyLength;

    return List.generate(_specialCharsKeyRows.length, (int rowNum) {
      List<SecureKeyboardKey> rowKeys = [];

      switch (rowNum) {
        case 3:
          // Shift
          rowKeys.add(_shiftActionKey());

          rowKeys.addAll(_getStringKeyRow(_specialCharsKeyRows, rowNum));

          emptyLength = _specialCharsKeyRows[rowNum].length;
          randomIndex = random.nextInt(emptyLength) + 1;
          rowKeys.insert(randomIndex, _blankActionKey());

          // Backspace
          rowKeys.add(_backspaceActionKey());
          break;
        case 4:
          // SpecialChars
          rowKeys.add(_specialCharsActionKey());

          // Clear
          rowKeys.add(_clearActionKey());

          // Confirm
          rowKeys.add(_confirmActionKey());
          break;
        default:
          rowKeys = _getStringKeyRow(_specialCharsKeyRows, rowNum);
          emptyLength = _specialCharsKeyRowMaxLength - _specialCharsKeyRows[rowNum].length;

          for (var i=0; i<emptyLength; i++) {
            randomIndex = random.nextInt(_specialCharsKeyRowMaxLength);

            if (randomIndex == _specialCharsKeyRowMaxLength - 1)
              rowKeys.add(_blankActionKey());
            else
              rowKeys.insert(randomIndex, _blankActionKey());
          }
      }

      return rowKeys;
    });
  }

  /// Returns the string type key row.
  List<SecureKeyboardKey> _getStringKeyRow(List<List<String>> keyRows, int rowNum) {
    return List.generate(keyRows[rowNum].length, (int keyNum) {
      String key = keyRows[rowNum][keyNum];

      return SecureKeyboardKey(
        text: key,
        capsText: key.toUpperCase(),
        type: SecureKeyboardKeyType.String
      );
    });
  }

  /// Returns the backspace action key.
  SecureKeyboardKey _backspaceActionKey() {
    return SecureKeyboardKey(
      type: SecureKeyboardKeyType.Action,
      action: SecureKeyboardKeyAction.Backspace
    );
  }

  /// Returns the confirm action key.
  SecureKeyboardKey _confirmActionKey() {
    return SecureKeyboardKey(
      type: SecureKeyboardKeyType.Action,
      action: SecureKeyboardKeyAction.Confirm
    );
  }

  /// Returns the clear action key.
  SecureKeyboardKey _clearActionKey() {
    return SecureKeyboardKey(
      type: SecureKeyboardKeyType.Action,
      action: SecureKeyboardKeyAction.Clear
    );
  }

  /// Returns the shift action key.
  SecureKeyboardKey _shiftActionKey() {
    return SecureKeyboardKey(
      type: SecureKeyboardKeyType.Action,
      action: SecureKeyboardKeyAction.Shift
    );
  }

  /// Returns the blank action key.
  SecureKeyboardKey _blankActionKey() {
    return SecureKeyboardKey(
      type: SecureKeyboardKeyType.Action,
      action: SecureKeyboardKeyAction.Blank
    );
  }

  /// Returns the specialChars action key.
  SecureKeyboardKey _specialCharsActionKey() {
    return SecureKeyboardKey(
      type: SecureKeyboardKeyType.Action,
      action: SecureKeyboardKeyAction.SpecialChars
    );
  }
}

example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:flutter_secure_keyboard/flutter_secure_keyboard.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      debugShowCheckedModeBanner: false,
      home: WithSecureKeyboardExample()
    );
  }
}

class WithSecureKeyboardExample extends StatefulWidget {
  @override
  _WithSecureKeyboardExampleState createState() => _WithSecureKeyboardExampleState();
}

class _WithSecureKeyboardExampleState extends State<WithSecureKeyboardExample> {
  final secureKeyboardController = SecureKeyboardController();

  final passwordEditor = TextEditingController();
  final passwordTextFieldFocusNode = FocusNode();

  final pinCodeEditor = TextEditingController();
  final pinCodeTextFieldFocusNode = FocusNode();

  @override
  Widget build(BuildContext context) {
    return WithSecureKeyboard(
      controller: secureKeyboardController,
      child: Scaffold(
        appBar: AppBar(title: Text('with_secure_keyboard_example')),
        body: _buildContentView()
      ),
    );
  }

  Widget _buildContentView() {
    // We recommend using the ListView widget to prevent widget overflows.
    return ListView(
      padding: const EdgeInsets.all(8.0),
      children: [
        Padding(
          padding: const EdgeInsets.only(bottom: 12.0),
          child: _buildPasswordTextField()
        ),
        _buildPinCodeTextField()
      ],
    );
  }

  Widget _buildPasswordTextField() {
    return Column(
      crossAxisAlignment: CrossAxisAlignment.start,
      children: [
        Text('Password'),
        TextFormField(
          controller: passwordEditor,
          focusNode: passwordTextFieldFocusNode,
          // We recommended to set false to prevent the soft keyboard from opening.
          enableInteractiveSelection: false,
          obscureText: true,
          onTap: () {
            secureKeyboardController.show(
              type: SecureKeyboardType.Alphanumeric,
              textFieldFocusNode: passwordTextFieldFocusNode,
              initText: passwordEditor.text,
              hintText: 'password',
              // Use onCharCodesChanged to have text entered in real time.
              onCharCodesChanged: (List<int> charCodes) {
                passwordEditor.text = String.fromCharCodes(charCodes);
              }
            );
          },
        ),
      ],
    );
  }

  Widget _buildPinCodeTextField() {
    return Column(
      crossAxisAlignment: CrossAxisAlignment.start,
      children: [
        Text('PinCode'),
        TextFormField(
          controller: pinCodeEditor,
          focusNode: pinCodeTextFieldFocusNode,
          // We recommended to set false to prevent the soft keyboard from opening.
          enableInteractiveSelection: false,
          obscureText: true,
          onTap: () {
            secureKeyboardController.show(
              type: SecureKeyboardType.Numeric,
              textFieldFocusNode: pinCodeTextFieldFocusNode,
              initText: pinCodeEditor.text,
              hintText: 'pinCode',
              // Use onDoneKeyPressed to allow text to be entered when you press the done key,
              // or to do something like encryption.
              onDoneKeyPressed: (List<int> charCodes) {
                pinCodeEditor.text = String.fromCharCodes(charCodes);
              }
            );
          },
        ),
      ],
    );
  }
}